flink-runtime/src/main/java/org/apache/flink/runtime/dispatcher/Dispatcher.java

```diff
@@ -1513,13 +1517,20 @@ private CompletableFuture<Void> waitForTerminatingJob(
                                                     throwable));
                                 });
 
-        return jobManagerTerminationFuture.thenAcceptAsync(
+        // keep track of the job as outstanding, if not done
+        if (!jobManagerTerminationFuture.isDone()) {
+            submittedAndWaitingTerminationJobIDs.add(jobId);
+        }
```

Comment:
This doesn't seem safe; there's no guarantee the termination future doesn't complete a millisecond after the condition, so any submission with the same job ID could still fail.

Reply:
Moved to the `internalSubmitJob` to always add before returning with a `waitForTerminatingJob` future for removing a possible race. It's now removed in `handleTermination`